src/invert/laplace/impls/petscamg/petscamg.cxx

```diff
@@ -0,0 +1,241 @@
+  //  else opts=opt;
+  opts = Options::getRoot()->getSection("petscamg");
+  opts->get("rtol",rtol,pow(10.0,-11),true);
+  opts->get("atol",atol,pow(10.0,-25),true);
+  opts->get("dtol",dtol,pow(10.0,5),true);
+  opts->get("maxits",maxits,200000,true);
+  opts->get("rightpre",rightpre,0,true);
+  opts->get("smtype",mgsm,1,true);
+  opts->get("jacomega",omega,0.8,true);
+  opts->get("checking",fcheck,1,true);
+  opts->get("multigridlevel",mglevel,6,true);
+  opts->get("solvertype",soltype,"gmres");
+
+  // Initialize, allocate memory, etc.
+  comms_tagbase = 385; // Some random number
+  
+  int implemented_global_flags = INVERT_START_NEW;
+  if ( global_flags & ~implemented_global_flags ) {
+    throw BoutException("Attempted to set Laplacian inversion flag that is not implemented in LaplaceMultigrid.");
+  }
+  int implemented_boundary_flags = INVERT_AC_GRAD + INVERT_SET + INVERT_DC_GRAD; // INVERT_DC_GRAD does not actually do anything, but harmless to set while comparing to Fourier solver with Neumann boundary conditions
+  if ( inner_boundary_flags & ~implemented_boundary_flags ) {
+    throw BoutException("Attempted to set Laplacian inner boundary inversion flag that is not implemented in LaplaceMultigrid.");
+  }
+  if ( outer_boundary_flags & ~implemented_boundary_flags ) {
+    throw BoutException("Attempted to set Laplacian outer boundary inversion flag that is not implemented in LaplaceMultigrid.");
+  }
+  if (nonuniform) {
+    throw BoutException("nonuniform option is not implemented in LaplaceMultigrid.");
+  }
+  
+  commX = localmesh->getXcomm();    // Where to get This mesh //
+
+  MPI_Comm_size(commX,&xNP);
+  MPI_Comm_rank(commX,&xProcI); 
+  Nx_local = localmesh->xend - localmesh->xstart + 1; // excluding guard cells
+  Nx_global = localmesh->GlobalNx - 2*localmesh->xstart; // excluding guard cells
+  mxstart = localmesh->xstart;
+  
+  if (mgcount == 0) {
+    output <<"Nx="<<Nx_global<<"("<<Nx_local<<")"<<endl;
+  }
+  zNP = 1;
+  zProcI = 0;
+  Nz_global = localmesh->GlobalNz;
+  Nz_local = Nz_global;
+  mzstart = 0; //  
+  // No parallelization in z-direction (for now)
+  // 
+  //else {
+  //  Nz_local = localmesh->zend - localmesh->zstart + 1; // excluding guard cells
+  //  Nz_global = localmesh->GlobalNz - 2*localmesh->zstart; // excluding guard cells
+  // }
+  if (mgcount==0) {
+    output <<"Nz="<<Nz_global<<"("<<Nz_local<<")"<<endl;
+  }
+
+  int ig,jg,ll,lg;
+  // Periodic boundary condition for z-direction
+  // Nz_global = 0
+  //
+  
+  nxt = Nx_local;
+  nzt = Nz_local;
+  xgstart = xProcI*nxt;
+  zgstart = zProcI*nzt;
+  lxs = 0;
+  lzs = 0;
+  if(xNP > 1) {
+    if(xProcI > 0) {
+      lxs += 1;
+      nxt += 1;
+    }
+    if(xProcI < xNP -1) nxt += 1;
+  }
+  if(zNP > 1) {
+    nzt += 2;
+    lzs += 1;
+  }
+  Nlocal = nxt*nzt;
+  Nglobal = Nz_global*Nx_global;
+
+  if (mgcount == 0) {
+    output <<"NP="<<xNP<<"N="<<Nlocal<<"("<<Nglobal<<")"<<endl;
+  }
+  gindices = new int[Nlocal];
```

Comment:
This (the non-3D) solver was written by someone else who unfortunately tended to use more of a C-style approach as opposed to a C++ one. Cleaning it up could be quite a bit of work...

Reply:
Ah ok! If we can get sufficient tests on it, I'm happy to do the cleanup.

Reply:
There's an integrated test in `tests/integrated/test-petscamg-laplace`, but the test is failing at the moment for cases 3 & 4, which are Neumann and Dirichlet boundary conditions set to non-zero values. I don't know why and don't have time to dig at the moment... The `plotresults.py` script suggests the problem may be at the outer x-boundary rather than the inner one.